src/AppInstallerCommonCore/Archive.cpp

```diff
@@ -4,65 +4,43 @@
 
 namespace AppInstaller::Archive
 {
-    HRESULT ExtractArchive(const std::filesystem::path& archivePath, const std::filesystem::path& destPath)
+    using unique_pidlist_absolute = wil::unique_any<PIDLIST_ABSOLUTE, decltype(&::CoTaskMemFree), ::CoTaskMemFree>;
+    using unique_lpitemidlist = wil::unique_any<LPITEMIDLIST, decltype(&::CoTaskMemFree), ::CoTaskMemFree>;
+
+    HRESULT TryExtractArchive(const std::filesystem::path& archivePath, const std::filesystem::path& destPath)
     {
-        IFileOperation* pFileOperation;
-        HRESULT hr = CoCreateInstance(CLSID_FileOperation, NULL, CLSCTX_ALL, IID_PPV_ARGS(&pFileOperation));
+        wil::com_ptr<IFileOperation> pFileOperation;
+        RETURN_IF_FAILED(CoCreateInstance(CLSID_FileOperation, NULL, CLSCTX_ALL, IID_PPV_ARGS(&pFileOperation)));
+        RETURN_IF_FAILED(pFileOperation->SetOperationFlags(FOF_NO_UI));
+
+        wil::com_ptr<IShellItem> pShellItemTo;
+        RETURN_IF_FAILED(SHCreateItemFromParsingName(destPath.c_str(), NULL, IID_PPV_ARGS(&pShellItemTo)));
+
+        unique_pidlist_absolute pidlFull;
+        RETURN_IF_FAILED(SHParseDisplayName(archivePath.c_str(), NULL, &pidlFull, 0, NULL));
 
-        if (SUCCEEDED(hr))
+        wil::com_ptr<IShellFolder> pArchiveShellFolder;
+        RETURN_IF_FAILED(SHBindToObject(NULL, pidlFull.get(), NULL, IID_PPV_ARGS(&pArchiveShellFolder)));
+
+        wil::com_ptr<IEnumIDList> pEnumIdList;
+        RETURN_IF_FAILED(pArchiveShellFolder->EnumObjects(nullptr, SHCONTF_FOLDERS | SHCONTF_NONFOLDERS, &pEnumIdList));
+
+        unique_lpitemidlist pidlChild;
+        ULONG nFetched;
+        while (pEnumIdList->Next(1, wil::out_param_ptr<LPITEMIDLIST*>(pidlChild), &nFetched) == S_OK && nFetched == 1)
         {
-            hr = pFileOperation->SetOperationFlags(FOF_NO_UI);
-            if (SUCCEEDED(hr))
+            wil::com_ptr<IShellItem> pShellItemFrom;
+            STRRET strFolderName;
+            WCHAR szFolderName[MAX_PATH];

```

Comment:
Are the paths actually limited to `MAX_PATH`?  If not, I would prefer if we could do a variable length path so that we don't hit a length issue here.